src/node_file.h

```diff
@@ -17,45 +17,48 @@ using v8::Value;
 
 namespace fs {
 
-class FSReqWrap: public ReqWrap<uv_fs_t> {
+enum class Ownership { COPY, MOVE };
+
+class FSReqInfo {
  public:
-  enum Ownership { COPY, MOVE };
+  FSReqInfo(const char* syscall,
+            const char* data,
+            enum encoding encoding = UTF8) : encoding_(encoding),
+                                             syscall_(syscall),
+                                             data_(data) { }
 
-  inline static FSReqWrap* New(Environment* env,
-                               Local<Object> req,
-                               const char* syscall,
-                               const char* data = nullptr,
-                               enum encoding encoding = UTF8,
-                               Ownership ownership = COPY);
+  virtual ~FSReqInfo() { Release(); }
 
   inline void Dispose();
 
-  virtual void Reject(Local<Value> reject);
-  virtual void Resolve(Local<Value> value);
-
-  void ReleaseEarly() {
+  void Release() {
     if (data_ != inline_data()) {
       delete[] data_;
       data_ = nullptr;
     }
   }
 
+  enum encoding encoding() const { return encoding_; }
   const char* syscall() const { return syscall_; }
   const char* data() const { return data_; }
-  const enum encoding encoding_;
+  void SetData(const char* data) { data_ = data; }
 
-  size_t self_size() const override { return sizeof(*this); }
+  void* operator new(size_t size) = delete;
+  void* operator new(size_t size, char* storage) { return storage; }
+  char* inline_data() { return reinterpret_cast<char*>(this + 1); }
```

Comment:
Maybe add a `CHECK_EQ(data_, reinterpret_cast<char*>(this + 1))` here.

Reply:
good idea

Reply:
actually, I take that back. Adding that CHECK_EQ in here breaks things. I'll investigate further, but I think having the CHECK elsewhere may make more sense.

Reply:
I assume you mean it breaks because of the memcpy after the placement new?

Reply:
Yes